core/services/offchainreporting/OCRContract.go

```diff
@@ -0,0 +1,228 @@
+package offchainreporting
+
+import (
+	"context"
+	"math/big"
+	"strings"
+	"sync"
+	"time"
+
+	ethereum "github.com/ethereum/go-ethereum"
+	gethCommon "github.com/ethereum/go-ethereum/common"
+
+	"github.com/smartcontractkit/chainlink/core/services/eth"
+	"github.com/smartcontractkit/chainlink/core/store/models"
+	"github.com/smartcontractkit/chainlink/offchainreporting/confighelper"
+	"github.com/smartcontractkit/chainlink/offchainreporting/gethwrappers/offchainaggregator"
+	ocrtypes "github.com/smartcontractkit/chainlink/offchainreporting/types"
+
+	"github.com/ethereum/go-ethereum/accounts/abi"
+	"github.com/ethereum/go-ethereum/accounts/abi/bind"
+	"github.com/pkg/errors"
+)
+
+var (
+	OCRContractConfigSet = getConfigSetHash()
+)
+
+type (
+	LogConfigSet struct {
+		eth.GethRawLog
+		ConfigCount  uint64
+		Oracles      []gethCommon.Address
+		Transmitters []gethCommon.Address
+		Threshold    uint8
+		Config       []byte
+	}
+
+	OCRContract struct {
+		ethClient        eth.Client
+		configSetChan    chan offchainaggregator.OffchainAggregatorConfigSet
+		contractFilterer *offchainaggregator.OffchainAggregatorFilterer
+		contractCaller   *offchainaggregator.OffchainAggregatorCaller
+		contractAddress  gethCommon.Address
+		logBroadcaster   eth.LogBroadcaster
+
+		jobID       models.ID
+		transmitter Transmitter
+		contractABI abi.ABI
+	}
+
+	Transmitter interface {
+		CreateEthTransaction(ctx context.Context, toAddress gethCommon.Address, payload []byte) error
+		FromAddress() gethCommon.Address
+	}
+)
+
+var (
+	_ ocrtypes.ContractConfigTracker      = &OCRContract{}
+	_ ocrtypes.ContractTransmitter        = &OCRContract{}
+	_ ocrtypes.ContractConfigSubscription = &OCRContractConfigSubscription{}
+)
+
+func NewOCRContract(address gethCommon.Address, ethClient eth.Client, logBroadcaster eth.LogBroadcaster, jobID models.ID, transmitter Transmitter) (o *OCRContract, err error) {
+	contractFilterer, err := offchainaggregator.NewOffchainAggregatorFilterer(address, ethClient)
+	if err != nil {
+		return o, errors.Wrap(err, "could not instantiate NewOffchainAggregatorFilterer")
+	}
+
+	contractCaller, err := offchainaggregator.NewOffchainAggregatorCaller(address, ethClient)
+	if err != nil {
+		return o, errors.Wrap(err, "could not instantiate NewOffchainAggregatorCaller")
+	}
+
+	contractABI, err := abi.JSON(strings.NewReader(offchainaggregator.OffchainAggregatorABI))
+	if err != nil {
+		return o, errors.Wrap(err, "could not get contract ABI JSON")
+	}
+
+	return &OCRContract{
```

Comment:
I personally am a fan of not using named fields in these literals since the compiler will then error if you forget to initialize something.

Reply:
One downside, which is that if you have two fields of the same type next to one another, you can easily mix them up.

Reply:
Agreed, it's a trade off. Too bad that go makes us choose 😦